### PR TITLE
UpnpExtraHeaders changes

### PR DESCRIPTION
**Pull request by AlaricSenat**
This set of commits change the way libupnp exposes HTTP headers. basically it exposes all the HTTP headers of the incoming message.

The motivations behind this change are explained in the commit messages.

For the context I'm implementing an UPNP/DLNA server module for the VLC media player using libupnp. The fact that the UpnpExtraHeaders interface is not forward compatible friendly is a bit bothering us.

For the moment the changes are not tested, I plan to add some unit tests eventually, maybe I should wait for GTest to be fully integrated the repo ? The PR is there so early mostly to get your opinion on the changes.

### CODE REVIEW

**Comment by Vollstrecker** [comment]
If you're able to write a gtest, feel free to do so. For now this part is deactivated because the only gtest isn't ported to the new interface. If you add a working one, just activate the part for for inclusion, add you test in gtest/ and comment out the old one. If I didn't miss something, you should be done. If additional args are to be added, feel free to extend the function for that or tell me what you're missing.

**Comment by mrjimenez** [comment]
The renaming part is maybe a point to question, I don't know. Lets hear what people have to say about that.

Other than that, I don't see any problem with your idea, I could commit it right now.

**Comment by AlaricSenat** [comment]
> The renaming part is maybe a point to question, I don't know. Lets hear what people have to say about that.

I agree that renaming the API calls feels kinda big for the actual benefit of these changes...

However if the changes (without renaming) are kept as they are, `UpnpExtraHeader` would feel incoherent as a name, because it wouldn't represent the "extra unrecognized headers" anymore. So I still think renaming is necessary if we change the behavior of the `UpnpExtraHeader` that way!

Maybe we could find a way to mark `UpnpExtraHeader` deprecated for now?

**Comment by mrjimenez** [comment]
I see no problem because the interface has been broken already. I just wanted to see if anyone would complain.

**Comment by whyman** [comment]
Does this also mean we can override any header by appending items? Or will it send both I suppose?

**Comment by mrjimenez** [comment]
The way it is currently done, I believe it will send both. But it is up to you what you want to do with them.

I think this patch only gives the outside world access to more information. Previously the known headers (ACCEPT, CONTENT-TYPE, etc.) were not exposed.

**Comment by whyman** [comment]
I think its a nice change, and leaves room for improvement. If there is going to be a major API bump soon then it makes sense to put this in too.

File: upnp/src/genlib/net/http/httpparser.c [hunk 1]
```diff
@@ -2221,53 +2221,41 @@ const char *method_to_str(http_method_t method)
...
-                                        extraHeader);
-                        UpnpListInsert(
-                                list, UpnpListEnd(list), extraHeaderNode);
-                        UpnpExtraHeaders_strncpy_name(extraHeader,
-                                header->name.buf,
-                                header->name.length);
-                        UpnpExtraHeaders_strncpy_value(extraHeader,
-                                header->value.buf,
-                                header->value.length);
+                extraHeader = UpnpHttpHeaders_new();
```
[the remark this answers is not in the document]
Comment (mrjimenez) [reply upnp/src/genlib/net/http/httpparser.c:2247]: Yep. :+1:

**Comment by AlaricSenat** [comment]
Thanks for the review !

Next step is unit test and once I'm done and sure that everything is stable, I'll set the PR as mergeable.

**Comment by Vollstrecker** [comment]
Interesting. About the tests and the headers I can't say that much, so I focus on what know about.

- The Macro you introduce looks like just copy&paste from [here](https://cliutils.gitlab.io/modern-cmake/chapters/testing/googletest.html) without respecting the hints about package should not be used literally. You even copied the typo in the comment.

- Then in general, why a new macro? In cmake/test-functions.cmake is a function that is meant to solve this all in one place. This one intentionally uses cmake_parse_arguments to be extensible to provide such stuff as linking against more than just gmock.

- You link against upnp-stable but you set all the include-dirs the static lib already sets. You've found the "modern-cmake" article and do this the vintage way. In fact you've missed  ${PUPNP_BINARY_DIR}/upnp/inc/ where upnpconfig.h is created and it only works because the lib tells your test where to search. The threadutil-dir is an exception from this, but the first line of this  file is meant as an example and covers exactly this.

- You're linking against gtest gmock gtest_main. This will fail if the googletest found on your system is used, as this targets are not created by module shipped with cmake. As all distro-maintainers I know won't enable downloading of external stuff if it's not absolutely needed (and even then they consider this a bug) this will fail there. So you should use GTest::gmock there. This ony depends on GTest::gtest, so this one hasn't to be listed on this.

- You use gtest_discover_tests to add the tests. This works as you only use upnp-static. If you want to test the dynamic one (I always prefer to test all we have) this will fail at least on Win, as you are not able to set the right path to find the libs.

- You set the VS_DEBUGGER... explcitely to the default that is used anyway.

- You set FOLDER property. That would be needed if you process actual test-data and the test i designed to find them at a specific location. Beside that I don't know if you do that (at least you don't add some), and I always prefer to tell the test where to find it's data. It would be easier to read that data from the actual WORKING_DIR which your set to somewhere else, and you set that folder to one that doesn't even exist (again this was copied).

- Sidenote, as this was copied, I hardly doubt that gtest_main is needed at all.

**Comment by Vollstrecker** [comment]
P.S. As there are many guesses and insinuation in my comment, I just gave it a try.

If you just change https://github.com/pupnp/pupnp/blob/951f1c4f11d3975d29935d42d98947ec77f5f3b7/gtest/CMakeLists.txt#L3 to be 
```
addGMock (test_UpnpHttpHeaderList test_UpnpHttpHeaderList.cpp
```
it just works.

**Comment by AlaricSenat** [comment]
Cool, was about to ask some opinions on the CMake part as I'm a complete newbie on that domain. Thanks for the quick feedback, I'll clean that up.

By the way, I just read through #271. Can you confirm me that Google test is the way you plan to do unit-testing in the future ? I noticed that there already a bunch of tests written for CTest. If you don't plan to port them on GoogleTest at all it might just be better that I write mine with CTest, having two unit testing framework feels like it add unnecessary complexity to me.

**Comment by Vollstrecker** [comment]
In #271 someone wanted to change the complete project-philisophy, but yes, the good outcome was that gtest is now easy useable. For the CTest, googletest uses this to, as this is just the cmake-part for running tests in a defined way. If you use google-test, or unit-test (which is not a framework at all, just a bunch of programs that run and return something), doesn't matter. It's even possible to write code that doesn't compile and just checks if an assertion we expect is thrown while compilation.

In fact I just don't care what is tested and how it is implemented. The more usefull tests we have the better.

**Comment by mrjimenez** [comment]
Hi folks,

I'd just like to add that @AlaricSenat should not wait for code or infra-structure to be added. Finish your pull request and once you decide it's fine we can move forward. If that is the case, we will improve it later.

**Comment by AlaricSenat** [comment]
I think I'm mostly done then!
The CI keeps timing out for OSX for some reasons tho...

**Comment by Vollstrecker** [comment]
I'm not sure if the change of the function name is the best thing, as it unconditionally adds gmock to all (that's why it was named that way), but as unconditional adding is something I always critisise, I need to take care of this after the holidays.

So I am happy with current state.

**Comment by mrjimenez** [comment]
Hi @AlaricSenat ,

> I think I'm mostly done then!
> The CI keeps timing out for OSX for some reasons tho...

I saw that and I thought it was suspicious. I did a clone of your changes and merged it by hand, and the results are here #296 . As you can see, passes all testes.

If you authorize me, I can merge it from there. As you can also see, your commits are preserved.

Regards,
Marcelo.

**Comment by AlaricSenat** [comment]
> I'm not sure if the change of the function name is the best thing, as it unconditionally adds gmock to all (that's why it was named that way), but as unconditional adding is something I always critisise, I need to take care of this after the holidays.

I agree. I thought of adding another argument to `addGTest` (former `addGMock`) to link with either `GTest::gmock` or `GTest::gtest` based on the value of the argument. I can take care of it if you are ok with that.

> If you authorize me, I can merge it from there. As you can also see, your commits are preserved.

If you don't have anything against  unnecessary merge commits on master, no problem!

**Comment by Vollstrecker** [comment]
It seems like something changed since I tried. Should we give Win and Mac a shot,too?

**Comment by Vollstrecker** [comment]
>I agree. I thought of adding another argument to addGTest (former addGMock) to link with either GTest::gmock or GTest::gtest based on the value of the argument. I can take care of it if you are ok with that solution.

Nothing different than what I would do. This or just the lib to link against directly, as you can see in the add_unit_test (or what I called it, didn't check.

**Comment by mrjimenez** [comment]
I'll close this since the PR has been accepted in #296

**Pull request closed without merging**